corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js

```diff
@@ -441,37 +455,46 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
             var self = this;
             e.preventDefault();
 
-            // validateFields will likely already have been called when user blurred the last field,
-            // but call it here just in case they didn't fill anything out
-            self.validateFields().done(function () {
+            self.validateAllFields().done(function () {
                 FormplayerFrontend.trigger("menu:query", self.getAnswers());
             });
         },
 
+        validateFieldChange: function(changedChildView) {
```

Comment:
Open to suggestion if there's a way to DRY validateFieldChange and validateAllFields since they are very similar

Reply:
This is tough, it's right on the border (if it were obvious I suspect you'd have done it already :) ). You could do a callback-heavy helper function like this:
```
_validateOnFormplayer: function (childViewIsValid, updateUI) {
            var self = this;
            var promise = $.Deferred();
            var invalidFields = [];

            self._updateModelsForValidation().done(function (response) {
                // Gather error messages
                self.children.each(function (childView) {
                    if (childViewIsValid(childView)) {
                        invalidFields.push(childView.model.get('text'));
                    }
                });

                updateUI(invalidFields);

                // incorporating other comment about resolve vs reject
               promise.resolve(response);
            });

            return promise;
},

validateFieldChange: function () {
    this._validateOnFormplayer(function (childView) {
         return (!childView.hasRequiredError() || childView === changedChildView) && !childView.isValid());
    }, function (invalidFields) {
          // nothing to do
    });
},

validateAllFields: function () {
    this._validateOnFormplayer(function (childView) {
          return !childView.isValid();
    }, function (invalidFields) {
           FormplayerFrontend.trigger('clearNotifications');
           ....
    });
},
...
```
...but that's just something to think about, it's debatable whether it simplifies the code. (also the naming & conditionals in there are rough)

Reply:
Appreciate the alternative approach - it's an interesting pattern for me to keep in mind. In this case, I do think it's a bit harder to follow with the callbacks so I'm thinking to keep it as is for now.